dashboard/lib/middlewares/redis_session_store.rb

```diff
@@ -0,0 +1,38 @@
+# frozen_string_literal: true
+
+module Middlewares
+  class RedisSessionStore < ActionDispatch::Session::RedisStore
+    # Extends +ActionDispatch::Request::Session+ to allow identifying whether a session has been changed.
+    module SessionExtension
+      def changed?
+        @changed.present?
+      end
+
+      private def load_for_write!
+        @changed = true
+        super
+      end
+    end
+
+    def initialize(...)
+      ActionDispatch::Request::Session.prepend(SessionExtension)
+      super
+    end
```

Comment:
I'm not 100% sure how I feel about doing a `prepend` in a class method initialization rather than in a Rails initializer. On the one hand, I like that it means this class is entirely self-contained; on the other hand, I worry about the potential for complications during the Rails initialization process or about the possibility that multiple instances of the session store could result in multiple prependings.

Do you have examples of this pattern of a prepend inside an initialize being successfully used elsewhere? Right now both in our codebase and in the Rails codebase itself, I think that prepending is only ever done as part of the Rails initialization process.

Reply:
> I worry about the potential for complications during the Rails initialization process or about the possibility that multiple instances of the session store could result in multiple prependings.

Ruby will ignore any additional attempts to prepend the same module, so there’s no need to worry about it:
``` ruby
class MyClass; end

module MyModule; end

MyClass.prepend(MyModule)
MyClass.prepend(MyModule)
```

```bash
irb(main):017:0> MyClass.ancestors
=> [MyModule, MyClass, Object, PP::ObjectMixin, Kernel, BasicObject]
```

Reply:
The Rails codebase is rich with examples of metaprogramming, as it's one of Ruby’s core strengths.
runtimehttps://github.com/rails/rails/blob/v6.1.7.7/activesupport/lib/active_support/fork_tracker.rb#L44-L49

Reply:
The Rails codebase is rich with examples of metaprogramming, as it's one of Ruby’s core strengths.
https://github.com/rails/rails/blob/v6.1.7.7/activesupport/lib/active_support/fork_tracker.rb#L44-L49

Reply:
However, I don’t see any issue with moving this part to a separate initializer either

Reply:
That's comforting, thank you! Both example and link are good context. I think I'm okay with this as-is, then; I still don't _love_ the idea of the prepend happening later in the initialization process than it otherwise would, but I think the advantages from a code organization perspective are enough to outweigh that. Thanks!